### ISSUE TITLE
FEM error

### ISSUE THREAD

**Issue by jc-olalla** [LoLab-VU/PyDREAM#18]
Dear all,

I am using pydream to optimize a function that uses a Finite Element Method (FEM) model as forward operator. Before using pydream with the FEM operator, I tested it with a much simpler analytical solution. It works well.

The problem that I encounter when running the FEM code is that the parameters are becoming negative. However, that should be impossible since I am working with the logarithm of the parameters. So, my wild guess is that the forward operators are somehow interfering with each other.

Has anyone experienced this sort of issue? 

Kind regards,

Juan

**Comment by ortega2247** [comment]
Hi Juan. Thank you for reporting this issue. Unfortunately, I don't have experience with FEM but I could try to see what's going on. What distributions did you use to set your priors? How does your likelihood function look like? If you send us a reproducible example of your issue, we could try to debug it.

Thanks for using pydream

**Comment by jc-olalla** [comment]
Thanks for the quick reply Oscar. I find pydream quite handy.

I am thinking of ways to upload a reproducible example. The problem is that the forward model uses two other python libraries that you likely don't have (pygimli and pybert). Or do you?

**My prior is a uniform distribution:**

`lower_limits = np.ones_like(original_params)`
`lower_limits[9:] = 10.0`
`upper_limits = 10 * np.ones_like(lower_limits)`
`upper_limits[9:] = 80.0`
`scale = np.log10(upper_limits) - np.log10(lower_limits)`
`parameters_to_sample = SampledParam(uniform, loc=np.log10(lower_limits), scale=scale)`

**The likelihood function follows the same structure as that in the robertson_nopysb example:**

`def likelihood(parameter_vector):`
`    parameter_vector = 10**np.array(parameter_vector)`
`    print('Parameter vector')`
`    # Solve fop`
`    npar = len(parameter_vector)`
`    y0 = pg.RVector(npar)  # array-like` 
`    for i in range(npar):`
`        y0[i] = parameter_vector[i]`

 `   cout = fop.response(y0)`

`    # Calculate log probability contribution given simulated experimental values.`
`    logp_ctotal = np.sum(like_ctot.logpdf(cout))`
    
`    # If simulation failed due to integrator errors, return a log probability of -inf.`
`    if np.isnan(logp_ctotal):`
`        logp_ctotal = -np.inf`
      
`    return logp_ctotal`

Juan

**Comment by ortega2247** [comment]
I suspect that the problem is related to the way pydream reflects sampled parameter values that are out of the boundaries defined by the uniform distributions. As it was mentioned in issue #9 pydream currently only reflects these values once, but this reflection could still be out of the uniform distribution boundaries. I am going to try and fix this issue this week to see if that fixes your problem.

Also, I can install pygimli and pybert to try and run your pydream calibration for debugging purposes.

**Comment by ortega2247** [comment]
Also, the easiest way to check if that is the problem is to use a distribution with soft boundaries. Could you please try to run your calibration with a normal distribution or other soft boundary distribution to see if the algorithm stops sampling negative parameter values?

**Comment by jc-olalla** [comment]
I ran1000 iterations with a normal distribution at different scales:

`parameters_to_sample = SampledParam(norm, loc=original_params, scale=scale/20.0)`
`parameters_to_sample = SampledParam(norm, loc=original_params, scale=scale/10.0)`

"original_params" is the log10 of the best estimate I had obtained from a conjugate gradient approach.
"scale" is the scale used in the uniform distribution

When using a narrow prior (scale/20.0), the algorithm does not sample negative values and gets to the 1000th iteration
When using a wider prior (scale/10.0),  the algorithm reached the 970th iteration, then it sampled negative values and stopped.

**Comment by ortega2247** [comment]
I see. This is unexpected. I double-checked the code and it also shouldn't be sampling negative values from the normal distribution. Which pydream version are you using? In the first post you mention said "So, my wild guess is that the forward operators are somehow interfering with each other." is there a reason why you think that's happening?

**Comment by jc-olalla** [comment]
I figured the problem out. It has nothing to do with pydream. There is an interpolating function within my model which sometimes produces negative values. This happens for a given combination of sampled model parameters. 

It happens often when the parameter space is wide. It didn't happen when I ran other optimization algorithms. So, I didn't know that my model had that problem. In conclusion, I am making my model more robust in order to run it in pydream.

Well, sorry for the inconvenience and thanks for quickly replying. That helped me figure out the problem. 

Best regards,

Juan

**Comment by jc-olalla** [comment]
I figured the problem out. It has nothing to do with pydream. There is an interpolating function within my model which sometimes produces negative values. This happens for a given combination of sampled model parameters. 

It happens often when the parameter space is wide. It didn't happen when I ran other optimization algorithms. So, I didn't know that my model had that problem. In conclusion, I am making my model more robust in order to run it in pydream.

Well, sorry for the inconvenience and thanks for quickly replying. That helped me figure out the problem. 

Best regards,

Juan
